### PR TITLE
Migrate to nom 7

### PR DESCRIPTION
**Pull request by mstange**
This passes tests.

I haven't yet fed a full xul.sym to into it. This is probably a good opportunity to try out socc-pair.

### CODE REVIEW

File: breakpad-symbols/src/sym_file/parser.rs [hunk 1]
```diff
@@ -390,23 +415,16 @@ impl SymbolParser {
...
                     // conservatively reject the entire parse.
                     return Err(SymbolError::ParseError("failed to parse file", self.lines));
                 }
-                Incomplete(_) => {
-                    // One of our sub-parsers wants more input, which normally
-                    // would be fine for a streaming parser, bust the newline
-                    // preprocessing we do means this should never happen.
-                    // So Incomplete input is just another kind of parsing Error.
-                    return Err(SymbolError::ParseError("line was incomplete!", self.lines));
-                }
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:409]: good call. The Incomplete stuff was an artifact of me trying to use the streaming machinery and then realizing it didn't actually work for this problem.

File: breakpad-symbols/src/sym_file/parser.rs [hunk 2]
```diff
@@ -27,265 +35,282 @@ enum Line {
...
-              add_rules: Default::default(),
-          }
-      }
-));
+fn stack_cfi_init(input: &[u8]) -> IResult<&[u8], StackInfoCfi> {
+    let (input, (_, address, size, rules)) = tuple((
+        terminated(tag("STACK CFI INIT"), multispace1),
+        terminated(hex_str_u64, multispace1),
+        terminated(hex_u32, multispace1),
+        terminated(map_res(not_my_eol, str::from_utf8), my_eol),
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:286]: I Can't Believe It's Not EOL!™️

File: breakpad-symbols/src/sym_file/parser.rs [hunk 3]
```diff
@@ -27,265 +35,282 @@ enum Line {
...
-              line,
-          }
-      }
-));
+fn func_line_data(input: &[u8]) -> IResult<&[u8], SourceLine> {
+    let (input, (address, size, line, file)) = tuple((
+        terminated(hex_str_u64, multispace1),
+        terminated(hex_u32, multispace1),
+        terminated(decimal_u32, multispace1),
+        terminated(decimal_u32, my_eol),
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:149]: I believe you're preserving semantics but it's interesting that we'll freak out if there's trailing whitespace

File: breakpad-symbols/src/sym_file/parser.rs [hunk 4]
```diff
@@ -27,265 +35,282 @@ enum Line {
...
-              address,
-              parameter_size,
-              name: name.to_string()
-          }
-      }
-));
+fn public_line(input: &[u8]) -> IResult<&[u8], PublicSymbol> {
+    let (input, (_, _multiple, _, address, parameter_size, name)) = tuple((
+        tag("PUBLIC"),
+        opt(tag(" m")),
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:127]: i believe the existing code allowed for arbitrary spaces here?

File: breakpad-symbols/src/sym_file/parser.rs [hunk 5]
```diff
@@ -27,265 +35,282 @@ enum Line {
...
-));
+fn module_line(input: &[u8]) -> IResult<&[u8], ()> {
+    let (input, _) = tuple((
+        terminated(tag("MODULE"), multispace1),
+        terminated(non_space, multispace1),  // os
+        terminated(non_space, multispace1),  // cpu
+        terminated(hex_digit1, multispace1), // debug id
+        terminated(not_my_eol, my_eol),      // filename
+    ))(input)?;
+    Ok((input, ()))
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:92]: huh, interesting that we discard this, I never even registered that these fields existed in a symbol file.

File: breakpad-symbols/src/sym_file/parser.rs [hunk 6]
```diff
@@ -1,8 +1,16 @@
...
-use nom::IResult::*;
-use nom::*;
+use nom::branch::alt;
+use nom::bytes::complete::{tag, take_while};
+use nom::character::complete::{char, digit1, hex_digit1, multispace1};
+use nom::character::{is_digit, is_hex_digit};
+use nom::combinator::{map, map_res, opt};
+use nom::error::{Error, ErrorKind, ParseError};
+use nom::multi::many0;
+use nom::number::complete::hex_u32;
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:11]: hey nom, why the fuck do you have this function for exactly and only u32

File: breakpad-symbols/src/sym_file/parser.rs [hunk 7]
```diff
@@ -27,265 +35,282 @@ enum Line {
...
+fn my_eol(input: &[u8]) -> IResult<&[u8], char> {
+    preceded(many0(char('\r')), char('\n'))(input)
+}
 
-// Match a decimal string, parse it to a u32.
-named!(decimal_u32<&[u8], u32>, map_res!(map_res!(digit, str::from_utf8), FromStr::from_str));
+/// Accept everything except `\r` and `\n`.
+///
+/// This is different from `not_line_ending` which rejects its input if it's
+/// followed by a `\r` which is not immediately followed by a `\n`.
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:66]: god do we actually rely on this, windows why

**Review by Gankra — APPROVED** [review]
looks solid, but yeah tossing it at socc-pair a bit seems like A Good Idea.

These days when I use modern nom I tend to do a bit more of an imperative style just because the adaptor soup gets headachey, but this works and makes the review very easy, so I'm totally happy with it.

**Comment by Gankra** [comment]
**None of the following is needed or requested**, errors for this parser aren't even user-facing. I've just been in a deep hole this month trying to figure out how to get good parsing errors, and wanted to spread the good word in case it's ever useful.

One little tip that I didn't understand until recently: when you have an `alt` your errors turn to garbage because nom can't distinguish between "didn't match this pattern" and "did match the pattern but the subpattern was corrupt". You can use [cut](https://docs.rs/nom/latest/nom/combinator/fn.cut.html) to specify "at this point, I am confident I have matched the pattern that an outer alt is checking for, so if the parse fails then the input is corrupt and this is to blame".

So basically if you toss a `cut` after each line's initial `tag` you can properly detect which line variant failed to parse. You can further use [the context combinator and convert_error](https://github.com/Geal/nom/blob/main/doc/error_management.md) to add better descriptions of what went wrong in the parse.

File: breakpad-symbols/src/sym_file/parser.rs (hunk 5 above)
Comment (luser) [reply breakpad-symbols/src/sym_file/parser.rs:92]: I guess it could be nice to get the debug ID and filename out to validate that they match what you expect, but AFAIK nothing has ever done that and it hasn't seemed to matter. IME, the other bits are mostly useful just for manually spot-checking symbol files.

**Comment by luser** [comment]
Nice work! I don't know if the tests here are complete enough to rely on, but we do have fuzz targets in this crate as well, so for additional confidence you could try `cargo fuzz`ing it.

File: breakpad-symbols/src/sym_file/parser.rs (hunk 1 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/parser.rs:409]: Oh I wasn't thinking any smart thoughts about this, I was just forced into this change by the type system, because nom's "complete" parsers can no longer return an Incomplete status.

File: breakpad-symbols/src/sym_file/parser.rs (hunk 6 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/parser.rs:11]: I know, right!?

**Comment by mstange** [comment]
I fixed the `m` space issue and added some calls to `cut` but I didn't add any nice error messages.

And I fed [this xul.sym](https://symbols.mozilla.org/xul.pdb/2E7FA147D9CC4B434C4C44205044422E1/xul.sym) into it and it worked.

I also found out that using socc-pair is a lot less fun if you don't have crash-stats minidump access or a bunch of existing minidumps on your machine.

**Comment by Gankra** [comment]
> I also found out that using socc-pair is a lot less fun if you don't have crash-stats minidump access or a bunch of existing minidumps on your machine.

Oh whoops, I thought that permission was super normal to have. Is it not?

**Comment by mstange** [comment]
Maybe it is, I'm not sure. I don't spend a lot of time debugging crashes so I've never asked for access. Anyway, Jeff downloaded a dump for me and I'm currently hacking up socc-pair so that I can compare a local minidump with a socorro JSON.

**Comment by mstange** [comment]
The socc-pair run passed, so I think this is good to go now. :)

**Comment by gabrielesvelto** [comment]
> Oh whoops, I thought that permission was super normal to have. Is it not?

It's rather uncommon because of the toxic nature of the data. @mstange do request protected data acccess, it will make your life a lot easier and given you've been doing a lot of work in this area it's definitely something you need. The instructions are [here](https://crash-stats.mozilla.org/documentation/protected_data_access/).

**Comment by Gankra** [comment]
Benchmark results:

# Summary

runtime:
* **cached: 1762ms -> 2084ms (oof!)**
* http: 6482ms -> 6590ms

memory usage unchanged

# methodology

```
> socc-pair --bench=5  --run-local=<snip> --api-token=<snip> --crash-id=d0170905-001e-430d-a664-452cc0220630 --mock-server --output-file=results-main-serverless.txt
> socc-pair --bench=5  --run-local=<snip> --api-token=<snip> --crash-id=d0170905-001e-430d-a664-452cc0220630 --clean-cache --output-file=results-main-download.txt

> git checkout nom7  

> socc-pair --bench=5  --run-local=<snip> --api-token=<snip> --crash-id=d0170905-001e-430d-a664-452cc0220630 --clean-cache --output-file=results-nom7-download.txt
> socc-pair --bench=5 --run-local=<snip> --api-token=<snip> --crash-id=d0170905-001e-430d-a664-452cc0220630 --mock-server --output-file=results-nom7-serverless.txt
```


# main, serverless (--mock-server)

```
benchmark results (ms):
  1750, 1773, 1798, 1758, 1735, 
average runtime: 00m:01s:762ms (1762ms)
median runtime: 00m:01s:758ms (1758ms)
min runtime: 00m:01s:735ms (1735ms)
max runtime: 00m:01s:798ms (1798ms)

max memory (rss) results (bytes):
  260980736, 260538368, 261054464, 260374528, 260497408, 
average max-memory: 248MB (260689100 bytes)
median max-memory: 248MB (260538368 bytes)
min max-memory: 248MB (260374528 bytes)
max max-memory: 248MB (261054464 bytes)
```

# nom7, serverless (--mock-server)

```
benchmark results (ms):
  2146, 2093, 2077, 2087, 2021, 
average runtime: 00m:02s:084ms (2084ms)
median runtime: 00m:02s:087ms (2087ms)
min runtime: 00m:02s:021ms (2021ms)
max runtime: 00m:02s:146ms (2146ms)

max memory (rss) results (bytes):
  259751936, 260878336, 260886528, 260718592, 260202496, 
average max-memory: 248MB (260487577 bytes)
median max-memory: 248MB (260718592 bytes)
min max-memory: 247MB (259751936 bytes)
max max-memory: 248MB (260886528 bytes)
```

# main, download (--clean-cache)

```
benchmark results (ms):
  6855, 6562, 6264, 6357, 6375, 
average runtime: 00m:06s:482ms (6482ms)
median runtime: 00m:06s:375ms (6375ms)
min runtime: 00m:06s:264ms (6264ms)
max runtime: 00m:06s:855ms (6855ms)

max memory (rss) results (bytes):
  261070848, 259944448, 260919296, 260550656, 260378624, 
average max-memory: 248MB (260572774 bytes)
median max-memory: 248MB (260550656 bytes)
min max-memory: 247MB (259944448 bytes)
max max-memory: 248MB (261070848 bytes)
```

# nom7, download (--clean-cache)

```
benchmark results (ms):
  6504, 6724, 6665, 6503, 6555, 
average runtime: 00m:06s:590ms (6590ms)
median runtime: 00m:06s:555ms (6555ms)
min runtime: 00m:06s:503ms (6503ms)
max runtime: 00m:06s:724ms (6724ms)

max memory (rss) results (bytes):
  261120000, 261070848, 261193728, 261021696, 260825088, 
average max-memory: 248MB (261046272 bytes)
median max-memory: 248MB (261070848 bytes)
min max-memory: 248MB (260825088 bytes)
max max-memory: 249MB (261193728 bytes)
```

**Comment by Gankra** [comment]
I expected that benchmark to be a formality, but that's quite a regression. It's something we *could* swallow but I don't feel great about it given this isn't adding functionality or fixing some known bugs (?)

**Comment by Gankra** [comment]
Hmm ok actually I actually computed the multiplier and it's "just" 1.18x. Raw numbers seemed more dramatic. Still not *great* but maybe something we could eat since rust-minidump already halved these numbers over breakpad.

**Comment by mstange** [comment]
I've added some commits to optimize things. Tests still pass.

On my machine, the commits reduce the time from 991ms to 737ms, in the following steps:

991ms -> 901ms -> 856ms -> 752ms -> 737ms

**Comment by Gankra** [comment]
thanks a bunch!

**Comment by mstange** [comment]
Thank you too!
